plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplitSource.java

```diff
@@ -178,74 +173,177 @@ public CompletableFuture<ConnectorSplitBatch> getNextBatch(ConnectorPartitionHan
+                            .includeColumnStats()
+                            .planFiles(),
+                    tableScan.targetSplitSize());
+            this.fileScanTaskIterator = fileScanTaskIterable.iterator();
+            this.next = getNext();
+        }
 
-            Schema fileSchema = scanTask.spec().schema();
-            Map<Integer, Optional<String>> partitionKeys = getPartitionKeys(scanTask);
+        @Override
+        public boolean hasNext()
+        {
+            return next != null;
+        }
 
-            Set<IcebergColumnHandle> identityPartitionColumns = partitionKeys.keySet().stream()
-                    .map(fieldId -> getColumnHandle(fileSchema.findField(fieldId), typeManager))
-                    .collect(toImmutableSet());
+        @Override
+        public ConnectorSplit next()
+        {
+            ConnectorSplit split = next;
+            next = getNext();
+            return split;
+        }
 
-            Supplier<Map<ColumnHandle, NullableValue>> partitionValues = memoize(() -> {
-                Map<ColumnHandle, NullableValue> bindings = new HashMap<>();
-                for (IcebergColumnHandle partitionColumn : identityPartitionColumns) {
-                    Object partitionValue = deserializePartitionValue(
-                            partitionColumn.getType(),
-                            partitionKeys.get(partitionColumn.getId()).orElse(null),
-                            partitionColumn.getName());
-                    NullableValue bindingValue = new NullableValue(partitionColumn.getType(), partitionValue);
-                    bindings.put(partitionColumn, bindingValue);
-                }
-                return bindings;
-            });
-
-            if (!dynamicFilterPredicate.isAll() && !dynamicFilterPredicate.equals(pushedDownDynamicFilterPredicate)) {
-                if (!partitionMatchesPredicate(
-                        identityPartitionColumns,
-                        partitionValues,
-                        dynamicFilterPredicate)) {
-                    continue;
-                }
-                if (!fileMatchesPredicate(
-                        fieldIdToType,
-                        dynamicFilterPredicate,
-                        scanTask.file().lowerBounds(),
-                        scanTask.file().upperBounds(),
-                        scanTask.file().nullValueCounts())) {
-                    continue;
-                }
+        @Override
+        public void close()
+                throws IOException
+        {
+            this.fileScanTaskIterable.close();
+            this.fileScanTaskIterator.close();
+        }
+
+        private IcebergSplit getNext()
+        {
+            if (!fileScanTaskIterator.hasNext()) {
+                return null;
+            }
+
+            return splitForFileScanTask(fileScanTaskIterator.next()).orElse(null);
+        }
+    }
+
+    private Optional<IcebergSplit> splitForFileScanTask(FileScanTask scanTask)
```

Comment:
Can you please extract some refactorings?

This shows up as a single huge change, but i guess most code is just being moved around.
Please make it apparent.